### PR TITLE
Big integers (Uint*) from byte slice array with `from_be_bytes`

### PR DESCRIPTION
**Pull request by dr-orlovsky**

### CODE REVIEW

**Comment by elichai** [comment]
Could you please implement this without heap allocations?
also, you can add a `slice_to_u64_be` here and use that: https://github.com/rust-bitcoin/rust-bitcoin/blob/f4e26caa9457348c62add2263407552efe9a7e21/src/util/endian.rs#L54

**Comment by dr-orlovsky** [comment]
> also, you can add a slice_to_u64_be here and use that:

@elichai I did, but it fails to compile since this function is not being used from anywhere (i.e. triggers `dead_code` error):

```console
error: function is never used: `slice_to_u64_be`
  --> src/util/endian.rs:4:9
   |
4  |         pub fn $name(slice: &[u8]) -> $type {
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
55 | define_slice_to_be!(slice_to_u64_be, u64);
   | ------------------------------------------ in this macro invocation
   |
note: the lint level is defined here
  --> src/lib.rs:43:9
   |
43 | #![deny(dead_code)]
   |         ^^^^^^^^^
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by elichai** [comment]
I meant that you'll use that instead of:
```rust

                let data: Vec<u64> = (0..$n_words)
                    .map(|n| {
                        let mut a = [0u8; 8];
                        a.copy_from_slice(&bytes[(n * 8)..(n * 8 + 8)]);
                        u64::from_be_bytes(a)
                    })
                    .collect();
```

**Comment by dr-orlovsky** [comment]
Oh yes, true

**Comment by dr-orlovsky** [comment]
done

**Comment by dr-orlovsky** [comment]
Now, with https://github.com/rust-bitcoin/rust-bitcoin/pull/435 merged, it will be nice to merge this one as well before the next version release

File: src/util/uint.rs [hunk 1]
```diff
@@ -73,18 +73,31 @@ macro_rules! construct_uint {
...
             }
 
+            /// Creates big integer value from a byte slice array using
+            /// big-endian encoding
+            pub fn from_be_bytes(bytes: [u8; $n_words * 8]) -> $name {
+                use super::endian::slice_to_u64_be;
+                let mut slice = [0u64; $n_words];
+                for n in 0..$n_words {
+                    slice[$n_words - n - 1] = slice_to_u64_be(&bytes[(n * 8)..(n * 8 + 8)]);
+                }
```
Comment (sgeisler) [inline src/util/uint.rs:97]: We could avoid all that hard to read byte arithmetic by using iterators.

```suggestion
                slice.iter_mut()
                    .rev()
                    .zip(bytes.chunks(8))
                    .for_each(|(word, bytes)| *word = slice_to_u64_be(bytes));
```

**Review by sgeisler — COMMENTED** [review]
Looks good. I think it could be more readable/idiomatic if it used iterators.

File: src/util/uint.rs (hunk 1 above)
Comment (elichai) [reply src/util/uint.rs:97]: (FWIW they result in the same ASM so perf wise they're the same: https://godbolt.org/z/9rnG9f)

Comment (Kixunil) [reply src/util/uint.rs:97]: Interesting, indexing is usually slower.

**Review by sgeisler — APPROVED** [review]

**Comment by dr-orlovsky** [comment]
Btw, trying to understand how these all is different from https://github.com/rust-bitcoin/bitcoin_hashes/blob/master/src/util.rs#L149-L210 and why not to move there all these functions (old and new ones)?

**Review by apoelstra — APPROVED** [review]
utACK

**Pull request merged**